polygeist/tools/cgeist/Lib/CodeGenTypes.cc

```diff
@@ -43,6 +43,11 @@ static cl::opt<bool>
 
 static cl::opt<bool> memRefABI("memref-abi", cl::init(true),
                                cl::desc("Use memrefs when possible"));
+// Remove this option when _Float16 promotion to float before operating works
+// for all cases.
+static llvm::cl::opt<bool>
+    Float16Support("ffloat16", llvm::cl::init(false),
```

Comment:
No need for a flag, that's dictated by `TargetInfo::hasFloat16Type()`

Reply:
Targets not having such type first promote to `float` and then operate with the promoted value. We are currently not doing that, so I hid support for this type under that flag to avoid generating illegal code by default.

Reply:
I dropped the flag and added tests for targets needing and not needing the promotion. I used `shouldEmitFloat16WithExcessPrecision` instead of what you propose as that's the function which rules whether or not to extend to `float`.

Reply:
Shouldn't the FE have already changed the type to an FP32 on targets that do not have native support for FP16 ? 
Is that not happening?

Reply:
No fully, it emits a cast but codegen has then to handle the fp16 to/from float conversion via `llvm.convert.to.fp16` / `llvm.convert.from.fp16`.

But there is 2 types competing for this in clang `__fp16` and `_Float16` both lowering to llvm's `half` (under the right params). I forgot the details but in short, `__fp16` is normally a storage type but used by opencl c for representing `half`, `_Float16` is the C type and DPC++ uses it to implement `sycl::half`.

For us, and at this stage, it doesn't really matters as you can't really use `sycl::half` as storage because of a corner case.

Reply:
I'm confused by what clang does when -march=x86-64 (which from the lit test in the PR should not natively support half-precision floats). I was expecting that the LLVM IR would contain a `float` rather than a `half` type, however that is not the case: 
```
~/projects/intel-llvm  sycl-mlir ?  ./build/bin/clang++ ../tmp/float16.cpp -S -emit-llvm -march=x86-64                                                                                                                                                            
 ~/projects/intel-llvm  sycl-mlir ?  grep "foo" float16.ll                                                                                                                                                                                                         
define dso_local noundef half @_Z3fooDF16_(half noundef %f) #0 {
```

Reply:
Clang will extend each operand to operations involving `half` values to `float` beforehand and then truncate the result back to `half`.